[SPARK-17298][SQL] Require explicit CROSS join for cartesian products

srinathshankar opened the pull request:
## What changes were proposed in this pull request?

Require the use of CROSS join syntax in SQL (and a new crossJoin
DataFrame API) to specify explicit cartesian products between relations.
By cartesian product we mean a join between relations R and S where
there is no join condition involving columns from both R and S.

If a cartesian product is detected in the absence of an explicit CROSS
join, an error must be thrown. Turning on the
"spark.sql.crossJoin.enabled" configuration flag will disable this check
and allow cartesian products without an explicit CROSS join.

The new crossJoin DataFrame API must be used to specify explicit cross
joins. The existing join(DataFrame) method will produce a INNER join
that will require a subsequent join condition.
That is df1.join(df2) is equivalent to select \* from df1, df2.
## How was this patch tested?

Added cross-join.sql to the SQLQueryTestSuite to test the check for cartesian products. Added a couple of tests to the DataFrameJoinSuite to test the crossJoin API. Modified various other test suites to explicitly specify a cross join where an INNER join or a comma-separated list was previously used.

srinathshankar committed:
[SPARK-17298][SQL] Require explicit CROSS join for cartesian products

Require the use of CROSS join syntax in SQL (and a new crossJoin
DataFrame API) to specify explicit cartesian products between relations.
By cartesian product we mean a join between relations R and S where
there is no join condition involving columns from both R and S.

If a cartesian product is detected in the absence of an explicit CROSS
join, an error must be thrown. Turning on the
"spark.sql.crossJoin.enabled" configuration flag will disable this check
and allow cartesian products without an explicit CROSS join.

The new crossJoin DataFrame API must be used to specify explicit cross
joins. The existing join(DataFrame) method will produce a INNER join
that will require a subsequent join condition.
That is df1.join(df2) == select * from df1, df2.

srinathshankar commented:
@rxin @sameeragarwal @hvanhovell @davies @JoshRosen

SparkQA commented:
**[Test build #64590 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64590/consoleFull)** for PR 14866 at commit [`250d96d`](https://github.com/apache/spark/commit/250d96dcf43c09dbff821e5b20f3f0ea86f08887).

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1534,6 +1542,34 @@ object SimplifyCaseConversionExpressions extends Rule[LogicalPlan] {
+ * Throw an error if a cartesian product is found without an explicit cross join specified.
+ * This rule is effectively disabled if the CROSS_JOINS_ENABLED flag is true.
+ */
+case class CheckCartesianProducts(conf: CatalystConf) extends Rule[LogicalPlan] {
```
Shouldn't this be a part of CheckAnalysis?

srinathshankar replied:
As discussed, must be done after ReorderJoin. Added a comment explaining why.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1534,6 +1542,34 @@ object SimplifyCaseConversionExpressions extends Rule[LogicalPlan] {
+    if (conf.allowCartesianProduct) {
+      plan
+    } else plan transform {
+      case j @ Join(left, right, joinType, conditions)
```
`j @ Join(left, right, Inner(false) | LeftOuter | RightOuter | FullOuter, condition) =>`?

srinathshankar replied:
Done.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1534,6 +1542,34 @@ object SimplifyCaseConversionExpressions extends Rule[LogicalPlan] {
+    } else plan transform {
+      case j @ Join(left, right, joinType, conditions)
+      if Seq(Inner(false), LeftOuter, RightOuter, FullOuter).contains(joinType) =>
+        val hasJoinCondition =
```
move this to a function and call that function in the case guard.

srinathshankar replied:
Done.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -159,23 +159,30 @@ object ExtractEquiJoinKeys extends Logging with PredicateHelper {
+   * was involved in an explicit cross join. Also returns the entire list of join conditions for
+   * the left-deep tree.
+   */
+  def flattenJoin(plan: LogicalPlan, parentJoinType: Inner = Inner(false))
```
just pass in `explicitCartesian`?

srinathshankar replied:
Done.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1534,6 +1542,34 @@ object SimplifyCaseConversionExpressions extends Rule[LogicalPlan] {
 /**
+ * Check if ther eany cartesian products between joins of any type in the optimized plan tree.
+ * Throw an error if a cartesian product is found without an explicit cross join specified.
+ * This rule is effectively disabled if the CROSS_JOINS_ENABLED flag is true.
```
As discussed offline. Please document that we need to run this after join reordering.

srinathshankar replied:
Done.

SparkQA commented:
**[Test build #64590 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64590/consoleFull)** for PR 14866 at commit [`250d96d`](https://github.com/apache/spark/commit/250d96dcf43c09dbff821e5b20f3f0ea86f08887).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `case class CheckCartesianProducts(conf: CatalystConf) extends Rule[LogicalPlan] `
  * `case class Inner(explicitCartesian: Boolean) extends JoinType `

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64590/
Test FAILed.

SparkQA commented:
**[Test build #64612 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64612/consoleFull)** for PR 14866 at commit [`256ad85`](https://github.com/apache/spark/commit/256ad8583da6e97f5bdfc35bab0fb2734d0f9b18).

rxin commented on sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala:
```diff
@@ -642,7 +642,7 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
       """SELECT x.a, y.a, x.b, y.b
         |FROM (SELECT 1 AS a, 2 AS b) x
-        |INNER JOIN (SELECT 1 AS a, 2 AS b) y
+        |CROSS JOIN (SELECT 1 AS a, 2 AS b) y
```
this is not a cross join, is it?

srinathshankar replied:
The join condition is eliminated because it is "trivial" according to the foldable propagation rule. This is annoying, but an expected by-product of detecting joins in the optimization phase. I can check to see if there is a way to split up the operator optimization batch so the cartesian check can be inserted before the constant propagation.

SparkQA commented:
**[Test build #64612 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64612/consoleFull)** for PR 14866 at commit [`256ad85`](https://github.com/apache/spark/commit/256ad8583da6e97f5bdfc35bab0fb2734d0f9b18).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `case class CheckCartesianProducts(conf: CatalystConf)`

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64612/
Test FAILed.

SparkQA commented:
**[Test build #64667 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64667/consoleFull)** for PR 14866 at commit [`6085dea`](https://github.com/apache/spark/commit/6085deabf44a43df874edc0edcb379efe2d55226).

sameeragarwal commented on sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -698,6 +698,8 @@ private[sql] class SQLConf extends Serializable with CatalystConf with Logging {
 
   override def groupByOrdinal: Boolean = getConf(GROUP_BY_ORDINAL)
+
+  override def allowCartesianProduct: Boolean = getConf(CROSS_JOINS_ENABLED)
```
The `crossJoinEnabled` above also references the same underlying conf and is being used at multiple places in the code. Can we unify the two?

sameeragarwal replied:
Also, on a related note, do you think it'd make sense to rename this (and the underlying conf) to `requireExplicitCrossJoinSyntax` to better convey the implied semantics?

srinathshankar replied:
As discussed, I only have the crossJoinEnabled flag, which has been exposed in CatalystConf. Also as discussed, deprecating crossJoinEnabled and switching the conf flag to requireCrossJoinSyntax is a good idea, will address in a follow-up PR.

sameeragarwal commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala:
```diff
@@ -353,7 +353,7 @@ case class BroadcastNestedLoopJoinExec(
     val broadcastedRelation = broadcast.executeBroadcast[Array[InternalRow]]()
```
per offline discussion, depending on the semantics we decide, we should revisit whether we'd like to keep the physical plan check here L344

srinathshankar replied:
The physical plan checks have been removed.

SparkQA commented:
**[Test build #64667 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64667/consoleFull)** for PR 14866 at commit [`6085dea`](https://github.com/apache/spark/commit/6085deabf44a43df874edc0edcb379efe2d55226).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CheckCartesianProducts(conf: CatalystConf)`
  * `case class Inner(explicitCartesian: Boolean) extends JoinType `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64667/
Test FAILed.

sameeragarwal commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -933,6 +936,45 @@ object CombineLimits extends Rule[LogicalPlan] {
+   * Check if a join specified by left, right and condition is a cartesian product. Returns true if
+   * there are no join conditions involving references from both left and right.
+   */
+  def isCartesianProduct(left: LogicalPlan, right: LogicalPlan, condition: Option[Expression])
```
nit: might be cleaner to just pass in the `Join` node directly as a parameter

srinathshankar replied:
done.

sameeragarwal commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -933,6 +936,45 @@ object CombineLimits extends Rule[LogicalPlan] {
+               |${left.treeString(false).trim}
+               |and
+               |${right.treeString(false).trim}
+               |Use a CROSS JOIN to allow cartesian products between these relations""".stripMargin)
```
Should we also talk about the conf flag here?

srinathshankar replied:
done.

sameeragarwal commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala:
```diff
@@ -303,7 +303,7 @@ package object dsl {
       def join(
         otherPlan: LogicalPlan,
-        joinType: JoinType = Inner,
+        joinType: JoinType = Inner(false),
```
nit: not a big deal either way but explicitly naming the boolean parameters (`Inner(explicitCartesian = false)`) would be great for readability.

srinathshankar replied:
Now renamed.

sameeragarwal commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:
```diff
@@ -39,39 +38,44 @@ object ReorderJoin extends Rule[LogicalPlan] with PredicateHelper {
    * The joined plan are picked from left to right, prefer those has at least one join condition.
    *
-   * @param input a list of LogicalPlans to join.
+   * @param input a list LogicalPlans to inner join, and a bool speciying if an explicit cross join
```
nit: typo

srinathshankar replied:
fixed

srinathshankar committed:
Inner -> InnerLike, Inner(false) -> Inner, Inner(true) -> Cross

Remove extraneous allowCartesianProduct. Remove cross join check from physical ops

SparkQA commented:
**[Test build #64735 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64735/consoleFull)** for PR 14866 at commit [`166092b`](https://github.com/apache/spark/commit/166092b29752d1d27f89be6cac548a970afa30a9).

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -933,6 +936,47 @@ object CombineLimits extends Rule[LogicalPlan] {
+   * there are no join conditions involving references from both left and right.
+   */
+  def isCartesianProduct(join: Join)
+      : Boolean = {
```
nit: put boolean on the the same line as the def.

srinathshankar replied:
Done.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -933,6 +936,47 @@ object CombineLimits extends Rule[LogicalPlan] {
+ * the join between R and S is not a cartesian product and therefore should be allowed.
+ * The predicate R.r = S.s is not recognized as a join condition until the ReorderJoin rule.
+ */
+case class CheckCartesianProducts(conf: CatalystConf)
```
More general remark. Why is this part of optimization? We could for instance also trigger this check in `QueryExecution`.

srinathshankar replied:
This is just the earliest possible place to do the check. (In fact it ideally belongs in analysis). Could you elaborate on triggering this in QueryExecution ?

hvanhovell replied:
Yeah it is the earliest possible place. It is just not something I would expect from optimization.

If we are to add it in QueryExecution then I should do this after optimization and before planning:https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala#L76-L81

srinathshankar replied:
As discussed, while it is probably best for this to be in analysis (does throw an analysis exception after all), leaving it here until we want to reorganize the analysis/optimizer rules.

gatorsmile replied:
If we do it here, it does not cover all the cases, right? `Cartesian Product` might be chosen based on the statistics.

gatorsmile replied:
If we do it in the optimizer, it does not cover all the cases, right? Cartesian Product might be chosen based on the statistics.

gatorsmile replied:
If we do it in the optimizer, it does not cover all the cases, right? Cartesian Product might be chosen based on the statistics.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/joinTypes.scala:
```diff
@@ -28,6 +28,7 @@ object JoinType {
     case "leftanti" => LeftAnti
+    case "cross" => Cross
     case _ =>
       val supported = Seq(
```
also add cross to the list.

srinathshankar replied:
done.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:
```diff
@@ -39,39 +38,46 @@ object ReorderJoin extends Rule[LogicalPlan] with PredicateHelper {
+  def createOrderedJoin(input: Seq[(LogicalPlan, InnerLike)], conditions: Seq[Expression])
+    : LogicalPlan = {
     assert(input.size >= 2)
     if (input.size == 2) {
```
using a pattern match might be simpler here.

srinathshankar replied:
Sameer helped me interpret this comment. I *think* I've addressed this.

SparkQA commented:
**[Test build #64735 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64735/consoleFull)** for PR 14866 at commit [`166092b`](https://github.com/apache/spark/commit/166092b29752d1d27f89be6cac548a970afa30a9).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `sealed abstract class InnerLike extends JoinType `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64735/
Test FAILed.

srinathshankar committed:
Address comments and fix a python test

SparkQA commented:
**[Test build #64746 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64746/consoleFull)** for PR 14866 at commit [`d5113f3`](https://github.com/apache/spark/commit/d5113f33c012f58bb079474296fd6cef6f583b1f).

SparkQA commented:
**[Test build #3242 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3242/consoleFull)** for PR 14866 at commit [`d5113f3`](https://github.com/apache/spark/commit/d5113f33c012f58bb079474296fd6cef6f583b1f).

SparkQA commented:
**[Test build #64746 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64746/consoleFull)** for PR 14866 at commit [`d5113f3`](https://github.com/apache/spark/commit/d5113f33c012f58bb079474296fd6cef6f583b1f).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64746/
Test FAILed.

SparkQA commented:
**[Test build #3242 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3242/consoleFull)** for PR 14866 at commit [`d5113f3`](https://github.com/apache/spark/commit/d5113f33c012f58bb079474296fd6cef6f583b1f).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #64793 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64793/consoleFull)** for PR 14866 at commit [`82545f0`](https://github.com/apache/spark/commit/82545f0d3ac0b3af50c4249ef36a7515ffcdf718).

SparkQA commented:
**[Test build #64793 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64793/consoleFull)** for PR 14866 at commit [`82545f0`](https://github.com/apache/spark/commit/82545f0d3ac0b3af50c4249ef36a7515ffcdf718).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64793/
Test FAILed.

SparkQA commented:
**[Test build #64801 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64801/consoleFull)** for PR 14866 at commit [`3a84a8a`](https://github.com/apache/spark/commit/3a84a8ad61b70fc4efbe19fd1c9f383d003f42cf).

SparkQA commented:
**[Test build #64801 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64801/consoleFull)** for PR 14866 at commit [`3a84a8a`](https://github.com/apache/spark/commit/3a84a8ad61b70fc4efbe19fd1c9f383d003f42cf).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64801/
Test FAILed.

srinathshankar committed:
Fixed R tests to do cross join when there are no columns

SparkQA commented:
**[Test build #3243 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3243/consoleFull)** for PR 14866 at commit [`703a554`](https://github.com/apache/spark/commit/703a554a342d91fba3040a769a8f7ff786569f0c).

SparkQA commented:
**[Test build #64814 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64814/consoleFull)** for PR 14866 at commit [`703a554`](https://github.com/apache/spark/commit/703a554a342d91fba3040a769a8f7ff786569f0c).

SparkQA commented:
**[Test build #3243 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3243/consoleFull)** for PR 14866 at commit [`703a554`](https://github.com/apache/spark/commit/703a554a342d91fba3040a769a8f7ff786569f0c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #64814 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64814/consoleFull)** for PR 14866 at commit [`703a554`](https://github.com/apache/spark/commit/703a554a342d91fba3040a769a8f7ff786569f0c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64814/
Test PASSed.

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -2265,7 +2265,7 @@ setMethod("join",
+              sdf <- callJMethod(x@sdf, "crossJoin", y@sdf)
             } else {
               if (class(joinExpr) != "Column") stop("joinExpr must be a Column")
               if (is.null(joinType)) {
```
should the next line be "crossJoin" too?

srinathshankar replied:
I've added cross to allowable join types. With a joinExpr I think inner is appropriate

felixcheung replied:
could you add "cross" to the API doc - L2246

SparkQA commented:
**[Test build #64863 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64863/consoleFull)** for PR 14866 at commit [`7f3d67f`](https://github.com/apache/spark/commit/7f3d67fb6f4c49e14f67b4dda2e0e11e076808e5).

hvanhovell commented:
LGTM - pending jenkins.

srinathshankar commented:
I'll update the python and R APIs in a follow up. Right now in python and R a cross join is done if no join exprs/columns and join types are specified. It would be good to require explicit cross joins in these apis as well.

SparkQA commented:
**[Test build #64868 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64868/consoleFull)** for PR 14866 at commit [`686b549`](https://github.com/apache/spark/commit/686b54986875cd9d47d4b772764af06ba301d96e).

SparkQA commented:
**[Test build #64868 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64868/consoleFull)** for PR 14866 at commit [`686b549`](https://github.com/apache/spark/commit/686b54986875cd9d47d4b772764af06ba301d96e).
 * This patch **fails some tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64868/
Test FAILed.

SparkQA commented:
**[Test build #3245 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3245/consoleFull)** for PR 14866 at commit [`2509e45`](https://github.com/apache/spark/commit/2509e451326d673ba6ea9d4d9a4e3991ea73b291).

SparkQA commented:
**[Test build #64869 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64869/consoleFull)** for PR 14866 at commit [`2509e45`](https://github.com/apache/spark/commit/2509e451326d673ba6ea9d4d9a4e3991ea73b291).

SparkQA commented:
**[Test build #64863 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64863/consoleFull)** for PR 14866 at commit [`7f3d67f`](https://github.com/apache/spark/commit/7f3d67fb6f4c49e14f67b4dda2e0e11e076808e5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64863/
Test PASSed.

SparkQA commented:
**[Test build #3245 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3245/consoleFull)** for PR 14866 at commit [`2509e45`](https://github.com/apache/spark/commit/2509e451326d673ba6ea9d4d9a4e3991ea73b291).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #64869 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64869/consoleFull)** for PR 14866 at commit [`2509e45`](https://github.com/apache/spark/commit/2509e451326d673ba6ea9d4d9a4e3991ea73b291).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64869/
Test PASSed.

sameeragarwal commented:
LGTM

hvanhovell commented:
Merging to master. Thanks!

davies commented:
LGTM
